cap-primitives/src/std/fs/open.rs

```diff
@@ -0,0 +1,85 @@
+//! This defines `open`, the primary entrypoint to sandboxed `open`.
+
+#[cfg(debug_assertions)]
+use super::get_path;
+#[cfg(debug_assertions)]
+use crate::fs::{is_same_file, open_unchecked};
+use crate::fs::{open_impl, OpenOptions};
+use std::{fs, io, path::Path};
+
+/// Perform an `openat`-like operation, ensuring that the resolution of the path
+/// never escapes the directory tree rooted at `start`.
+pub fn open(start: &fs::File, path: &Path, options: &OpenOptions) -> io::Result<fs::File> {
+    // Call `open`.
+    let result = open_impl(start, path, options);
+
+    // Do an unsandboxed lookup and check that we found the same result.
+    // TODO: This is a racy check, though it is useful for testing and fuzzing.
+    #[cfg(debug_assertions)]
+    match open_unchecked(
```

Comment:
Will this even work in WASI?

Reply:
My plan for WASI is to have it skip this code altogether; details in an upcoming patch :-).